File: 1321-get-equal-substrings-within-budget/get-equal-substrings-within-budget.go
func equalSubstring(s string, t string, maxCost int) int {
    n := len(s)
    cost := make([]int,n)
    for i :=0 ; i< n ; i++ {
        cost[i] = abs(int(s[i]) - int(t[i]))
    }
    ans := 0 
    for i :=0 ; i< n ; {
        curr :=0
        j:=i
        for ; j < n ;  {
            curr += cost[j]
            if curr > maxCost {
                break
            }
            if (j-i +1) > ans {
                ans = (j-i + 1)
            }
            j+=1
        }
        if j== n {
            break
        }
        incr := true
        for ;curr > maxCost ; {
            curr -= cost[i]
            i+=1
            incr = false
        }
        if incr {
            i++
        }
    }
    return ans
}


func abs(a int) int {
    if a<0 {
        return -a
    }
    return a
}

File: 1321-get-equal-substrings-within-budget/get-equal-substrings-within-budget.py
class Solution:
    def equalSubstring(self, s: str, t: str, maxCost: int) -> int:
        n = len(s)
     
        ans = 0
        curr =0
        i = 0
        start= 0
        # sliding window
        while i<n:
            # print(i)
            curr += abs(ord(s[i]) - ord(t[i]))
            # move the back of the window forward
            while curr> maxCost :
                curr -= abs(ord(s[start]) - ord(t[start]))
                start+=1
            ans = max(ans,i-start+1)
            i+=1


        return ans
